dss/util/streaming.py

```diff
@@ -0,0 +1,111 @@
+"""
+Utilities in this module support the S3 chunked upload protocol as described in
```

Comment:
I actually think chunked uploads are really pivotal in letting us maximize the performance of this.

- Part size starts to increase past 64MB after the object exceeds 300GB in size, up to 500MB for a 5TB object (the max size for S3). 500MB is beyond the limit of what we should reasonably buffer. (I do need to add a TODO to properly segment parts beyond 300GB, though)
- Each lambda copies multiple parts at the same time to maximize the full-duplex use of its network interface and to amortize against TCP startup costs and occasional slow S3 connections. Using streaming lets us crank up that parallelism.
- I believe the real memory usage in Python is at least twice the size of the buffer.
- Lambdas don't have an actual filesystem. The FS is a ramdisk, so storing files on it will count against the available memory. But even if we did have an actual storage volume, I don't think it would be a good idea to write parts to it.
- After writing the code for the S3 streaming upload, and stress testing with files up to 100GB, I haven't observed any problems with it so far. I could add some debug logging in case we run into any in the future, though.

Just say no to bufferbloat 😁